jigna/js/src/app/async_proxy_factory.js

```diff
@@ -0,0 +1,182 @@
+/////////////////////////////////////////////////////////////////////////////
+// AsyncProxyFactory
+/////////////////////////////////////////////////////////////////////////////
+
+jigna._SavedData = function(data) {
+    // Used internally to save marshaled data to unmarshal later.
+    this.data = data;
+};
+
+jigna.AsyncProxyFactory = function(client) {
+    jigna.ProxyFactory.call(this, client);
+};
+
+jigna.AsyncProxyFactory.prototype = Object.create(
+    jigna.ProxyFactory.prototype
+);
+
+jigna.AsyncProxyFactory.prototype.constructor = jigna.AsyncProxyFactory
+
+jigna.AsyncProxyFactory.prototype._add_instance_attribute = function(proxy, attribute_name){
+    var descriptor, get, set;
+
+    get = function() {
+        // In here, 'this' refers to the proxy!
+        var value = this.__cache__[attribute_name];
+        if (value === undefined) {
+            value = this.__client__.get_attribute(this, attribute_name);
+            if (value === undefined) {
+                var info = this.__info__;
+                if (info && (info.attribute_values !== undefined)) {
+                    var index = info.attribute_names.indexOf(attribute_name);
+                    value = this.__client__._unmarshal(
+                        info.attribute_values[index]
+                    );
+                }
+            } else {
+                this.__cache__[attribute_name] = value;
+            }
+        }
+
+        return value;
+    };
+
+    set = function(value) {
+        // In here, 'this' refers to the proxy!
+        //
+        // If the proxy is for a 'HasTraits' instance then we don't need
+        // to set the cached value here as the value will get updated when
+        // we get the corresponding trait event. However, setting the value
+        // here means that we can create jigna UIs for non-traits objects - it
+        // just means we won't react to external changes to the model(s).
+        this.__cache__[attribute_name] = value;
+        this.__client__.set_instance_attribute(
+            this.__id__, attribute_name, value
+        );
+    };
+
+    descriptor = {enumerable:true, get:get, set:set, configurable:true};
+    Object.defineProperty(proxy, attribute_name, descriptor);
+};
+
+jigna.AsyncProxyFactory.prototype._populate_dict_proxy = function(proxy, info) {
+    var index, key;
+    var values = info.values;
+
+    for (index=0; index < info.keys.length; index++) {
+        key = info.keys[index];
+        this._add_item_attribute(proxy, key);
+        proxy.__cache__[key] = new jigna._SavedData(values.data[index]);
```

Comment:
The rationale is to not unpickle the data until the point you need it.  It is just an optimization.  Saved data just stores the json data and when a dictionary value is required the necessary object is constructed on demand instead of creating a million proxy objects that may not be used.